<a href="https://colab.research.google.com/github/zachfreitas/ADS_505_Applied_Data_Sci_for_Business-/blob/main/Module_2/Assignment_2/Assignment_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 2.1
Zachariah Freitas 2022-09-11

## Import required packages

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
pip install dmba

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 11.8 MB 5.2 MB/s 


In [51]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import matplotlib.pylab as plt

import dmba
from dmba import regressionSummary
from dmba import adjusted_r2_score, AIC_score, BIC_score


from sklearn.linear_model import LogisticRegressionCV

%matplotlib inline

In [13]:
# Get data
ts_df = pd.read_csv('https://raw.githubusercontent.com/zachfreitas/ADS_505_Applied_Data_Sci_for_Business-/main/Module_2/Assignment_2/Tayko.csv')
print(ts_df.shape)
print(ts_df.columns)
ts_df

(2000, 25)
Index(['sequence_number', 'US', 'source_a', 'source_c', 'source_b', 'source_d',
       'source_e', 'source_m', 'source_o', 'source_h', 'source_r', 'source_s',
       'source_t', 'source_u', 'source_p', 'source_x', 'source_w', 'Freq',
       'last_update_days_ago', '1st_update_days_ago', 'Web order',
       'Gender=male', 'Address_is_res', 'Purchase', 'Spending'],
      dtype='object')


sequence_number  US  source_a  source_c  source_b  source_d  source_e  \
0                   1   1         0         0         1         0         0   
1                   2   1         0         0         0         0         1   
2                   3   1         0         0         0         0         0   
3                   4   1         0         1         0         0         0   
4                   5   1         0         1         0         0         0   
...               ...  ..       ...       ...       ...       ...       ...   
1995             1996   1         0         0         0         0         0   
1996             1997   1         0         0         0         0         0   
1997             1998   1         0         0         0         0         0   
1998             1999   1         0         0         0         0         0   
1999             2000   1         0         0         0         0         0   

      source_m  source_o  source_h  ...  source_x  source_w  Freq  \
0            0         0         0  ...         0         0     2   
1            0         0         0  ...         0         0     0   
2            0         0         0  ...         0         0     2   
3            0         0         0  ...         0         0     1   
4            0         0         0  ...         0         0     1   
...        ...       ...       ...  ...       ...       ...   ...   
1995         0         0         0  ...         0         0     1   
1996         0         0         0  ...         0         0     1   
1997         0         0         0  ...         0         0     0   
1998         0         0         0  ...         0         1     1   
1999         0         0         0  ...         0         0     1   

      last_update_days_ago  1st_update_days_ago  Web order  Gender=male  \
0                     3662                 3662          1            0   
1                     2900                 2900          1            1   
2                     3883                 3914          0            0   
3                      829                  829          0            1   
4                      869                  869          0            0   
...                    ...                  ...        ...          ...   
1995                  1701                 1701          1            0   
1996                  2633                 2633          1            1   
1997                  3394                 3394          0            0   
1998                   253                  253          0            1   
1999                  1261                 1844          0            0   

      Address_is_res  Purchase  Spending  
0                  1         1       128  
1                  0         0         0  
2                  0         1       127  
3                  0         0         0  
4                  0         0         0  
...              ...       ...       ...  
1995               0         1        30  
1996               0         1        10  
1997               0         0         0  
1998               1         0         0  
1999               0         0         0  

[2000 rows x 25 columns]

## Question 1

Each catalog costs approximately $2 to mail (including printing, postage, and mailing costs). Estimate the gross profit that the firm could expect from the remaining 180,000 names if it selects them randomly from the pool. (6 Points)

In [31]:
# Calculate Gross Profit on 180k Names
historical_response_rate = 1065/200000 # Purchasers / Customer List - Approx:0.053
model_response_rate = 1000/2000 # Purchasers / (Purchasers + NonPurchasers) - Approx:0.5
probability_of_purchase = historical_response_rate/model_response_rate
list_size = 180000
potential_customers = probability_of_purchase * list_size
avg_cost = 2.0
avg_spend = ts_df.loc[ts_df['Purchase'] == 1]['Spending'].mean()

gross_profit = potential_customers * (avg_spend - avg_cost)
print(f'Gross Profit =${gross_profit:,.2f}')

Gross Profit =$389,628.33


## Question 2
Develop a model for classifying a customer as a purchaser or nonpurchaser

### Question 2.1
Partition the data randomly into a training set (800 records), validation set (700 records), and test set (500 records). (6 Points)


In [49]:
# Partition Data Set into Test Train.
trainData, testData = train_test_split(ts_df, test_size=0.25, random_state=1)
print('Train + Validation Data:',trainData.shape, 'Test Data:', testData.shape)

# The validation set is used to fine-tune the hyperparameters of the model and is considered a part of the training of the model. 
train_Data, valid_Data = train_test_split(trainData, test_size=(700/trainData.shape[0]), random_state=1)
print('Train Data:',train_Data.shape, 'Validation Data:', valid_Data.shape)

Train + Validation Data: (1500, 25) Test Data: (500, 25)
Train Data: (800, 25) Validation Data: (700, 25)


### Question 2.2
Run logistic regression with L2 penalty, using method LogisticRegressionCV, to
select the best subset of variables, then use this model to classify the data into purchasers and nonpurchasers. Use only the training set for running the model. (Logistic regression is used because it yields an estimated “probability of purchase,” which is required later in the analysis.) (6 Points)

In [62]:
# Data Review
ts_df.describe().T

count       mean          std  min      25%     50%  \
sequence_number       2000.0  1000.5000   577.494589  1.0   500.75  1000.5   
US                    2000.0     0.8245     0.380489  0.0     1.00     1.0   
source_a              2000.0     0.1265     0.332495  0.0     0.00     0.0   
source_c              2000.0     0.0560     0.229979  0.0     0.00     0.0   
source_b              2000.0     0.0600     0.237546  0.0     0.00     0.0   
source_d              2000.0     0.0415     0.199493  0.0     0.00     0.0   
source_e              2000.0     0.1510     0.358138  0.0     0.00     0.0   
source_m              2000.0     0.0165     0.127420  0.0     0.00     0.0   
source_o              2000.0     0.0335     0.179983  0.0     0.00     0.0   
source_h              2000.0     0.0525     0.223089  0.0     0.00     0.0   
source_r              2000.0     0.0685     0.252665  0.0     0.00     0.0   
source_s              2000.0     0.0470     0.211692  0.0     0.00     0.0   
source_t              2000.0     0.0215     0.145080  0.0     0.00     0.0   
source_u              2000.0     0.1190     0.323869  0.0     0.00     0.0   
source_p              2000.0     0.0060     0.077246  0.0     0.00     0.0   
source_x              2000.0     0.0180     0.132984  0.0     0.00     0.0   
source_w              2000.0     0.1375     0.344461  0.0     0.00     0.0   
Freq                  2000.0     1.4170     1.405738  0.0     1.00     1.0   
last_update_days_ago  2000.0  2155.1010  1141.302846  1.0  1133.00  2280.0   
1st_update_days_ago   2000.0  2435.6015  1077.872233  1.0  1671.25  2721.0   
Web order             2000.0     0.4260     0.494617  0.0     0.00     0.0   
Gender=male           2000.0     0.5245     0.499524  0.0     0.00     1.0   
Address_is_res        2000.0     0.2210     0.415024  0.0     0.00     0.0   
Purchase              2000.0     0.5000     0.500125  0.0     0.00     0.5   
Spending              2000.0   102.6250   186.782610  0.0     0.00     2.0   

                          75%     max  
sequence_number       1500.25  2000.0  
US                       1.00     1.0  
source_a                 0.00     1.0  
source_c                 0.00     1.0  
source_b                 0.00     1.0  
source_d                 0.00     1.0  
source_e                 0.00     1.0  
source_m                 0.00     1.0  
source_o                 0.00     1.0  
source_h                 0.00     1.0  
source_r                 0.00     1.0  
source_s                 0.00     1.0  
source_t                 0.00     1.0  
source_u                 0.00     1.0  
source_p                 0.00     1.0  
source_x                 0.00     1.0  
source_w                 0.00     1.0  
Freq                     2.00    15.0  
last_update_days_ago  3139.25  4188.0  
1st_update_days_ago   3353.00  4188.0  
Web order                1.00     1.0  
Gender=male              1.00     1.0  
Address_is_res           0.00     1.0  
Purchase                 1.00     1.0  
Spending               153.00  1500.0

In [63]:
ts_df.columns

Index(['sequence_number', 'US', 'source_a', 'source_c', 'source_b', 'source_d',
       'source_e', 'source_m', 'source_o', 'source_h', 'source_r', 'source_s',
       'source_t', 'source_u', 'source_p', 'source_x', 'source_w', 'Freq',
       'last_update_days_ago', '1st_update_days_ago', 'Web order',
       'Gender=male', 'Address_is_res', 'Purchase', 'Spending'],
      dtype='object')

In [69]:
from sklearn import preprocessing
# Prep data for Logistic Regression model
#scaler = preprocessing.StandardScaler()

scaler = preprocessing.MinMaxScaler()
# Fit the scaler in trianing dataset.
scaler.fit(train_Data[['1st_update_days_ago', 'last_update_days_ago', 'Freq']])  # Note the use of an array of column names

# Transform the full dataset
ts_df_Norm = pd.concat([ts_df[['sequence_number', 'US', 'source_a', 'source_c', 'source_b', 'source_d',
                              'source_e', 'source_m', 'source_o', 'source_h', 'source_r', 'source_s',
                              'source_t', 'source_u', 'source_p', 'source_x', 'source_w', 'Web order',
                              'Gender=male', 'Address_is_res', 'Purchase', 'Spending']],
                        pd.DataFrame(scaler.transform(ts_df[['1st_update_days_ago', 'last_update_days_ago', 'Freq']]), 
                                    columns=['1st_update_days_ago', 'last_update_days_ago', 'Freq'])], axis=1)

# Break-out the test and train datasets based on indexes
trainNorm = ts_df_Norm.iloc[train_Data.index]
display(trainNorm)

validNorm = ts_df_Norm.iloc[valid_Data.index]
display(validNorm)

testNorm = ts_df_Norm.iloc[testData.index]
display(testNorm)


sequence_number  US  source_a  source_c  source_b  source_d  source_e  \
746               747   1         0         1         0         0         0   
617               618   1         1         0         0         0         0   
1808             1809   1         0         0         1         0         0   
929               930   1         0         0         0         0         0   
878               879   1         0         0         1         0         0   
...               ...  ..       ...       ...       ...       ...       ...   
1127             1128   1         0         0         0         0         0   
1377             1378   1         0         0         1         0         0   
610               611   1         0         0         1         0         0   
1209             1210   1         0         0         1         0         0   
212               213   1         0         0         0         0         0   

      source_m  source_o  source_h  ...  source_x  source_w  Web order  \
746          0         0         0  ...         0         0          0   
617          0         0         0  ...         0         0          1   
1808         0         0         0  ...         0         0          1   
929          0         0         0  ...         0         1          0   
878          0         0         0  ...         0         0          0   
...        ...       ...       ...  ...       ...       ...        ...   
1127         0         0         0  ...         0         0          0   
1377         0         0         0  ...         0         0          1   
610          0         0         0  ...         0         0          0   
1209         0         0         0  ...         0         0          0   
212          0         0         0  ...         0         1          1   

      Gender=male  Address_is_res  Purchase  Spending  1st_update_days_ago  \
746             0               0         0         0             0.469027   
617             0               0         1      1195             0.941641   
1808            0               0         1       345             0.800765   
929             1               1         1       266             0.440564   
878             0               0         0         0             0.840708   
...           ...             ...       ...       ...                  ...   
1127            0               0         0         0             0.811528   
1377            1               0         0         0             0.873714   
610             0               0         0         0             0.846927   
1209            0               0         0         0             0.800287   
212             0               0         1       277             0.378378   

      last_update_days_ago      Freq  
746               0.469027  0.076923  
617               0.122698  0.923077  
1808              0.800526  0.153846  
929               0.114566  0.230769  
878               0.840708  0.230769  
...                    ...       ...  
1127              0.811528  0.076923  
1377              0.873714  0.076923  
610               0.846927  0.076923  
1209              0.800287  0.000000  
212               0.375508  0.153846  

[800 rows x 25 columns]

sequence_number  US  source_a  source_c  source_b  source_d  source_e  \
88                 89   0         0         0         0         0         0   
1925             1926   1         0         0         0         0         0   
1767             1768   1         0         0         0         0         0   
920               921   1         0         0         0         0         1   
1034             1035   1         0         0         0         0         1   
...               ...  ..       ...       ...       ...       ...       ...   
1057             1058   1         0         1         0         0         0   
1968             1969   1         0         0         0         1         0   
867               868   1         0         1         0         0         0   
482               483   0         0         0         0         0         0   
1969             1970   1         0         0         0         0         0   

      source_m  source_o  source_h  ...  source_x  source_w  Web order  \
88           0         0         0  ...         0         0          1   
1925         0         0         0  ...         0         1          0   
1767         0         0         0  ...         0         0          0   
920          0         0         0  ...         0         0          0   
1034         0         0         0  ...         0         0          0   
...        ...       ...       ...  ...       ...       ...        ...   
1057         0         0         0  ...         0         0          1   
1968         0         0         0  ...         0         0          0   
867          0         0         0  ...         0         0          1   
482          0         0         0  ...         0         1          0   
1969         0         0         0  ...         0         1          0   

      Gender=male  Address_is_res  Purchase  Spending  1st_update_days_ago  \
88              0               0         1       248             0.941641   
1925            1               0         1       163             0.169098   
1767            1               1         1       197             0.365702   
920             0               0         0         0             0.654150   
1034            0               0         1       361             0.572351   
...           ...             ...       ...       ...                  ...   
1057            0               1         0         0             0.469027   
1968            0               1         1       160             0.780914   
867             1               0         1       247             0.480507   
482             1               1         0         0             0.157379   
1969            1               0         1        81             0.296341   

      last_update_days_ago      Freq  
88                0.107869  0.153846  
1925              0.169098  0.076923  
1767              0.207606  0.307692  
920               0.654150  0.000000  
1034              0.126046  0.384615  
...                    ...       ...  
1057              0.467591  0.153846  
1968              0.676872  0.230769  
867               0.266443  0.230769  
482               0.157379  0.076923  
1969              0.293231  0.153846  

[700 rows x 25 columns]

sequence_number  US  source_a  source_c  source_b  source_d  source_e  \
674               675   1         0         0         0         0         0   
1699             1700   1         0         0         0         0         0   
1282             1283   1         1         0         0         0         0   
1315             1316   1         1         0         0         0         0   
1210             1211   0         0         0         0         0         1   
...               ...  ..       ...       ...       ...       ...       ...   
537               538   1         0         0         0         0         0   
1450             1451   1         1         0         0         0         0   
1919             1920   0         0         0         0         0         1   
255               256   1         0         0         0         0         0   
589               590   1         0         0         0         0         0   

      source_m  source_o  source_h  ...  source_x  source_w  Web order  \
674          0         0         1  ...         0         0          0   
1699         0         0         0  ...         0         1          1   
1282         0         0         0  ...         0         0          0   
1315         0         0         0  ...         0         0          1   
1210         0         0         0  ...         0         0          1   
...        ...       ...       ...  ...       ...       ...        ...   
537          0         0         0  ...         1         0          0   
1450         0         0         0  ...         0         0          1   
1919         0         0         0  ...         0         0          1   
255          0         0         1  ...         0         0          0   
589          0         0         0  ...         0         0          0   

      Gender=male  Address_is_res  Purchase  Spending  1st_update_days_ago  \
674             1               1         0         0             0.329825   
1699            0               0         1       184             0.109065   
1282            1               0         0         0             0.887348   
1315            1               0         1      1289             0.805788   
1210            0               0         0         0             0.659651   
...           ...             ...       ...       ...                  ...   
537             1               0         1        44             0.805788   
1450            1               0         1       281             0.505621   
1919            0               0         1       514             0.611576   
255             1               1         0         0             0.341784   
589             0               0         1        35             0.452523   

      last_update_days_ago      Freq  
674               0.320258  0.153846  
1699              0.102129  0.153846  
1282              0.887348  0.076923  
1315              0.268118  0.692308  
1210              0.659651  0.000000  
...                    ...       ...  
537               0.281512  0.076923  
1450              0.269792  0.153846  
1919              0.547477  0.230769  
255               0.336762  0.153846  
589               0.440804  0.153846  

[500 rows x 25 columns]

In [70]:
# Model: Logistic Regress to identify important variables.

from dmba.metric import AIC_score
from dmba import classificationSummary, gainsChart, liftChart

train_X = trainNorm.drop(['Purchase', 'Spending','sequence_number'], axis=1)
train_y = trainNorm['Purchase']
valid_X = validNorm.drop(['Purchase', 'Spending','sequence_number'], axis=1)
valid_y = validNorm['Purchase']

logit_rlr = LogisticRegressionCV(penalty="l2", solver='liblinear')
logit_rlr.fit(train_X, train_y)

print('regularization', logit_rlr.C_, '\n')
print('intercept ', logit_rlr.intercept_[0], '\n')

display(pd.DataFrame({'coeff': logit_rlr.coef_[0]}, index=train_X.columns).transpose())

print('\nAIC', AIC_score(valid_y, logit_rlr.predict(valid_X), df=len(train_X.columns) + 1), '\n')

# confusion matrix
classificationSummary(valid_y, logit_rlr.predict(valid_X))

regularization [21.5443469] 

intercept  -3.3054364245386894 



US  source_a  source_c  source_b  source_d  source_e  source_m  \
coeff  0.316449  2.066884 -0.228879  0.425102  1.530084  0.977436  1.379095   

       source_o  source_h  source_r  ...  source_u  source_p  source_x  \
coeff  0.572976  -4.19167  0.871888  ...   2.21869  2.446376  1.268546   

       source_w  Web order  Gender=male  Address_is_res  1st_update_days_ago  \
coeff  1.611026   0.752222    -0.321296        -0.58223            -0.600159   

       last_update_days_ago       Freq  
coeff              0.089199  23.691016  

[1 rows x 22 columns]


AIC 937.2823471519939 

Confusion Matrix (Accuracy 0.7914)

       Prediction
Actual   0   1
     0 295  78
     1  68 259


## Question 3

### Question 3.1

### Question 3.2

#### Question 3.2.1

#### Question 3.2.2

#### Question 3.2.3

## Question 4

### Question 4.1

### Question 4.2

### Question 4.3

### Question 4.4

### Question 4.5

### Question 4.6

## Question 5